src/docker.rs

```diff
@@ -325,10 +328,147 @@ pub fn run(
         }
     }
 
+    let mut image = image(config, target)?;
+
+    // TODO: if -p, use that
+    let package = metadata.packages.get(0).unwrap();
+
+    if let Some(dockerfile) = config.dockerfile(target)? {
+        image = Dockerfile::File {
+            path: &dockerfile,
+            context: config.dockerfile_context(target)?.as_deref(),
+            name: config.image(target)?.as_deref(),
+        }
+        .build(
+            verbose,
+            &host_root,
+            config.dockerfile_build_args(target)?.unwrap_or_default(),
+            target,
+            package,
+        )
+        .wrap_err("when building dockerfile")?;
+    }
+
+    if let Some(pre_build) = config.pre_build(target)? {
+        let stdin = Dockerfile::Stdin {
+            content: format!(
+                r#"
+FROM {image}
+ARG CROSS_CMD
+RUN eval "${{CROSS_CMD}}""#
+            ),
+        };
+        stdin
+            .build(
+                verbose,
+                &host_root,
+                Some(("CROSS_CMD", pre_build.join("\n"))),
+                target,
+                package,
+            )
+            .wrap_err("when pre-building")
+            .with_note(|| format!("CROSS_CMD={}", pre_build.join("\\n")))?;
+        image = stdin.image_name(target, package);
+    }
+
     docker
-        .arg(&image(config, target)?)
+        .arg(&*image)
         .args(&["sh", "-c", &format!("PATH=$PATH:/rust/bin {:?}", cmd)])
-        .run_and_get_status(verbose)
+        .run_and_get_status(verbose, false)
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Dockerfile<'a> {
+    File {
+        path: &'a str,
+        context: Option<&'a str>,
+        name: Option<&'a str>,
+    },
+    Stdin {
+        content: String,
+    },
+}
+
+impl<'a> Dockerfile<'a> {
+    pub fn build(
+        &self,
+        verbose: bool,
+        host_root: &Path,
+        build_args: impl IntoIterator<Item = (impl AsRef<str>, impl AsRef<str>)>,
+        target_triple: &Target,
+        package: &crate::cargo::Package,
+    ) -> Result<String> {
+        let mut docker_build = docker_command("build")?;
+        docker_build.current_dir(host_root);
+        docker_build.env("DOCKER_SCAN_SUGGEST", "false");
+        let image_name = self.image_name(target_triple, package);
+        docker_build.args(["--tag", &image_name]);
+
+        for (key, arg) in build_args.into_iter() {
+            docker_build.args(["--build-arg", &format!("{}={}", key.as_ref(), arg.as_ref())]);
+        }
+
+        match self {
+            Dockerfile::File { path, .. } => {
+                docker_build.args(["--file", path]);
+                if let Some(context) = self.context() {
+                    docker_build.arg(&context);
+                }
+            }
+            Dockerfile::Stdin { content } => {
+                // XXX: Ugly hack to circumvent platform issues. we could use echo, but it may not exist
+                let stdin_path = tempfile::NamedTempFile::new()
```

Comment:
I think we can have a better version: just write to a Dockerfile in the `/target/${target}` directory, and then it will automatically be cleaned up by `cargo clean`, and we don't have to worry about making 100s of tempfiles.

Reply:
yep, that's what I mentioned somewhere elsewhere, just slowly working through everything